File: Ass_6_11_2.py
import sys
from collections import deque

sys.setrecursionlimit(200000)

def bfs(start, target, n, adj):
    dist = [-1] * n
    queue = deque([start])
    dist[start] = 0
    
    while queue:
        curr = queue.popleft()
        
        # If we reach our destination, return the distance
        if curr == target:
            return dist[curr]
            
        for neighbor in adj[curr]:
            if dist[neighbor] == -1: # If unvisited
                dist[neighbor] = dist[curr] + 1
                queue.append(neighbor)
                
    return -1 # Path not found

def dfs(node, parent, visited, adj):
    visited[node] = True
    
    for neighbor in adj[node]:
        if not visited[neighbor]:
            if dfs(neighbor, node, visited, adj):
                return True
        # If visited and not the parent we just came from, it's a circle!
        elif neighbor != parent:
            return True
            
    return False

def solve():
    try:
        # Read straight from input.txt
        input_data = open("input.txt").read().split()
    except FileNotFoundError:
        print("Please create input.txt with your input data.")
        return

    if not input_data:
        return
        
    n = int(input_data[0])
    m = int(input_data[1])
    
    adj = [[] for _ in range(n)]
    
    idx = 2
    for _ in range(m):
        u = int(input_data[idx])
        v = int(input_data[idx+1])
        adj[u].append(v)

File: Ass_6_11_2.java
import java.util.*;

public class Ass_6_11_2 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        if (sc.hasNextInt()) {
            int n = sc.nextInt();
            int m = sc.nextInt();
            
            // Build the Graph
            List<List<Integer>> adj = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                adj.add(new ArrayList<>());
            }
            
            for (int i = 0; i < m; i++) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                adj.get(u).add(v);
                adj.get(v).add(u);
            }
            
            int s = sc.nextInt(); // Source section
            int d = sc.nextInt(); // Destination section
            
            // 1. Find shortest path using BFS
            int shortestPath = bfs(s, d, n, adj);
            System.out.println(shortestPath);
            
            // 2. Detect cycle using DFS
            boolean hasCycle = checkCycle(n, adj);
            if (hasCycle) {
                System.out.println("YES");
            } else {
                System.out.println("NO");
            }
        }
        sc.close();
    }
    
    // BFS implementation for Shortest Path
    static int bfs(int start, int target, int n, List<List<Integer>> adj) {
        int[] distance = new int[n];
        Arrays.fill(distance, -1); // -1 means unvisited
        Queue<Integer> queue = new LinkedList<>();
        
        queue.add(start);
        distance[start] = 0;
        
        while (!queue.isEmpty()) {
            int current = queue.poll();
            
            if (current == target) {
                return distance[current];
            }
            
            for (int neighbor : adj.get(current)) {
                if (distance[neighbor] == -1) { // If not visited
                    distance[neighbor] = distance[current] + 1;
                    queue.add(neighbor);
                }
            }
        }
        return -1; // Path not found
    }
    
    // Setup for DFS cycle check across all disconnected parts of the warehouse
    static boolean checkCycle(int n, List<List<Integer>> adj) {
        boolean[] visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                // -1 represents that the starting node has no parent
                if (dfs(i, -1, visited, adj)) {
                    return true; 
                }
            }
        }
        return false;
    }
    
    // DFS implementation for Cycle Detection
    static boolean dfs(int node, int parent, boolean[] visited, List<List<Integer>> adj) {
        visited[node] = true;
        
        for (int neighbor : adj.get(node)) {
            if (!visited[neighbor]) {
                if (dfs(neighbor, node, visited, adj)) {
                    return true;
                }
            } 
            // If the neighbor IS visited, but it's not the section we just came from...
            else if (neighbor != parent) {
                return true; // We found a cycle!
            }
        }
        return false;
    }
}
